internal/catalogmetadata/types.go

```diff
@@ -157,19 +157,15 @@ func (b *Bundle) HasDeprecation() bool {
 
 // IsDeprecated returns true if the bundle
 // has been explicitly deprecated. This can occur
-// in one of two ways:
-// - the olm.package the bundle belongs to has been deprecated
-// - the bundle itself has been deprecated
+// if the bundle itself has been deprecated.
 // this function does not take into consideration
-// olm.channel deprecations associated with the bundle
-// as a bundle can be present in multiple channels with
-// some channels being deprecated and some not.
+// olm.channel or olm.package deprecations associated
+// with the bundle as a bundle can be present in multiple
+// channels with some channels being deprecated and some not
+// Similarly if the package is deprecated then all the bundles
+// are deprecated and there should be no special treatment
 func (b *Bundle) IsDeprecated() bool {
 	for _, dep := range b.Deprecations {
-		if dep.Reference.Schema == declcfg.SchemaPackage && dep.Reference.Name == b.Package {
-			return true
-		}
-
```

Comment:
Removing this will actually mean that explicitly deprecated bundles will be preferred less than ones that aren't even if the entire package is deprecated - fixing

Reply:
Looking at this a little closer (sorry I didn't get the chance with the original PR), I'm a little concerned that we're putting package and channel level deprecations in the `Bundle` type.

I think I would have expected that we would put the package deprecation in the `Package` type, channel deprecation in the `Channel` type and bundle deprecation in the `Bundle` type.

And if we had that, we could use composition by embedding the deprecation type in each of the package/channel/bundle types and then use a shared `IsDeprecated` function that comes with that deprecation type.

Is it possible to do something like this?

Reply:
Probably, but my concern with doing that is that we would likely need to restructure resolution to provide all those types as results from the resolution. IIUC the resolution process currently will only return the `Bundle` type that contains all the necessary information for continuing with the install/upgrade process.

Reply:
Restructuring resolution to return variables for `Package`, `Channel`, and `Bundle` types is out of scope for this work IMO

Reply:
> Removing this will actually mean that explicitly deprecated bundles will be preferred less than ones that aren't even if the entire package is deprecated - fixing

One other question about this point. Think about this scenario:
- Package `foo` is deprecated
- Bundle `foo.v1.0.0` is not deprecated
- Bundle `foo.v1.0.1` is deprecated

I would expect `foo.v1.0.0` to be preferred over `foo.v1.0.1`. The `package` deprecation should not cascade to the bundles one way or the other.

Reply:
https://github.com/operator-framework/operator-controller/pull/577/commits/6fdc060493762e4d0e4c20577ffc415ac84645aa - reverts to where the check is removed and should satisfy the expectation you have.

Reply:
@joelanford Could you take another look when you have some time?